File: ocean/line_follow/host/activitybot_encoder_dashboard_raylib.c
#define _POSIX_C_SOURCE 200809L

#include <errno.h>
#include <fcntl.h>
#include <math.h>
#include <signal.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>

#include "raylib.h"

#ifndef SIM2REAL_ROOT
#define SIM2REAL_ROOT "."
#endif

#ifndef LINE_FOLLOW_ROOT
#define LINE_FOLLOW_ROOT SIM2REAL_ROOT
#endif

#define TRACE_CAP 4096
#define LOG_CAP 8
#define LOG_LEN 192
#define LINE_CAP 2048

typedef struct {
    double t;
    int left;
    int right;
    int left_raw;
    int right_raw;
} Sample;

typedef struct {
    float values[TRACE_CAP];
    int next;
    int count;
} Trace;

typedef struct {
    char lines[LOG_CAP][LOG_LEN];
    int next;
    int count;
} LogRing;

typedef struct {
    pid_t pid;
    int fd;
    int exit_code;
    bool exited;
    bool stopped_by_parent;
    char pending[LINE_CAP];
    int pending_len;
} LoaderProc;

typedef struct {
    const char *port;
    const char *board;
    const char *loader;
    const char *elf;
    const char *summary_file;
    int width;
    int height;
    double seconds;
} Args;

static const Color C_BG = {15, 18, 24, 255};
static const Color C_PANEL = {27, 31, 39, 255};
static const Color C_PANEL_DARK = {22, 25, 31, 255};
static const Color C_BAR_BG = {36, 39, 46, 255};
static const Color C_LINE_DIM = {80, 86, 96, 255};
static const Color C_GRID = {45, 50, 58, 255};
static const Color C_MUTED = {155, 164, 178, 255};
static const Color C_CYAN = {64, 196, 255, 255};
static const Color C_ORANGE = {255, 176, 64, 255};
static const Color C_GREEN = {94, 220, 126, 255};
static const Color C_OFF = {70, 75, 84, 255};

static double monotonic_seconds(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec + (double)ts.tv_nsec / 1000000000.0;
}

static void sleep_ms(long ms) {
    struct timespec req;
    req.tv_sec = ms / 1000;
    req.tv_nsec = (ms % 1000) * 1000000L;
    nanosleep(&req, NULL);
}

static const char *env_default(const char *name, const char *fallback) {
    const char *value = getenv(name);
    return (value && value[0]) ? value : fallback;
}

static void path_join(char *out, size_t out_size, const char *root, const char *suffix) {
    snprintf(out, out_size, "%s/%s", root, suffix);
}

static void trace_push(Trace *trace, float value) {
    trace->values[trace->next] = value;
    trace->next = (trace->next + 1) % TRACE_CAP;
    if (trace->count < TRACE_CAP) {
        trace->count++;
    }
}

static float trace_recent(const Trace *trace) {
    if (trace->count == 0) {
        return 0.0f;
    }
    int idx = trace->next - 1;
    if (idx < 0) {
        idx += TRACE_CAP;
    }
    return trace->values[idx];
}

static float trace_at_recent(const Trace *trace, int offset_from_oldest_recent, int recent_count) {
    int oldest = trace->next - trace->count;
    while (oldest < 0) {
        oldest += TRACE_CAP;
    }
    int skip = trace->count - recent_count;
    int idx = (oldest + skip + offset_from_oldest_recent) % TRACE_CAP;
    return trace->values[idx];
}

static void log_push(LogRing *logs, const char *line) {
    if (!line || !line[0]) {
        return;
    }
    snprintf(logs->lines[logs->next], LOG_LEN, "%s", line);
    logs->next = (logs->next + 1) % LOG_CAP;
    if (logs->count < LOG_CAP) {
        logs->count++;
    }
}

static const char *log_at(const LogRing *logs, int offset) {
    int oldest = logs->next - logs->count;
    while (oldest < 0) {
        oldest += LOG_CAP;
    }
    return logs->lines[(oldest + offset) % LOG_CAP];
}

static void draw_label(int x, int y, const char *text, int size, Color color) {
    DrawText(text, x, y, size, color);
}

static void draw_bar(int x, int y, int w, int h, float value, float peak, Color color) {
    DrawRectangle(x, y, w, h, C_BAR_BG);
    int fill = 0;
    if (peak > 0.0f) {
        fill = (int)fminf((float)w, fmaxf(0.0f, (value / peak) * (float)w));
    }
    DrawRectangle(x, y, fill, h, color);
    DrawRectangleLines(x, y, w, h, C_MUTED);
}

static void draw_trace(int x, int y, int w, int h, const Trace *trace, float peak, Color color) {
    DrawRectangle(x, y, w, h, C_PANEL_DARK);
    DrawRectangleLines(x, y, w, h, C_LINE_DIM);
    DrawLine(x, y + h / 2, x + w, y + h / 2, C_GRID);

    int recent = trace->count < w ? trace->count : w;
    if (recent < 2) {
        return;
    }

    float scale = fmaxf(peak, 1.0f);
    int start_x = x + (w - recent);
    float first = trace_at_recent(trace, 0, recent);
    int prev_x = start_x;
    int prev_y = y + h - (int)((first / scale) * (float)(h - 8)) - 4;
    if (prev_y < y + 3) prev_y = y + 3;
    if (prev_y > y + h - 3) prev_y = y + h - 3;

    for (int i = 1; i < recent; i++) {
        float value = trace_at_recent(trace, i, recent);
        int px = start_x + i;
        int py = y + h - (int)((value / scale) * (float)(h - 8)) - 4;
        if (py < y + 3) py = y + 3;
        if (py > y + h - 3) py = y + h - 3;
        DrawLine(prev_x, prev_y, px, py, color);
        prev_x = px;
        prev_y = py;
    }
}

static void strip_line(char *line) {
    char *dst = line;
    for (char *src = line; *src; src++) {
        if (*src != '\r') {
            *dst++ = *src;
        }
    }
    *dst = '\0';
    while (*line == ' ' || *line == '\t') {
        memmove(line, line + 1, strlen(line));
    }
    size_t n = strlen(line);
    while (n > 0 && (line[n - 1] == '\n' || line[n - 1] == ' ' || line[n - 1] == '\t')) {
        line[--n] = '\0';
    }
}

static bool parse_sample(const char *line, Sample *sample, double t) {
    int left = 0;
    int right = 0;
    int left_raw = 0;
    int right_raw = 0;

    const char *candidate = line;
    while (candidate && *candidate) {
        if (candidate[0] == 'E' && candidate[1] == ' ') {
            if (sscanf(candidate, "E %d %d %d %d", &left, &right, &left_raw, &right_raw) == 4) {
                sample->t = t;
                sample->left = left;
                sample->right = right;
                sample->left_raw = left_raw;
                sample->right_raw = right_raw;
                return true;
            }
        }
        candidate = strchr(candidate + 1, 'E');
    }
    return false;
}

static bool file_exists(const char *path) {
    return access(path, R_OK) == 0;
}

static int start_loader(LoaderProc *proc, const Args *args, LogRing *logs) {
    int pipefd[2];
    if (pipe(pipefd) != 0) {
        log_push(logs, strerror(errno));
        return -1;
    }

    pid_t pid = fork();
    if (pid < 0) {
        log_push(logs, strerror(errno));
        close(pipefd[0]);
        close(pipefd[1]);
        return -1;
    }

    if (pid == 0) {
        setpgid(0, 0);
        close(pipefd[0]);
        dup2(pipefd[1], STDOUT_FILENO);
        dup2(pipefd[1], STDERR_FILENO);
        close(pipefd[1]);
        execl(args->loader, "propeller-load", "-b", args->board, "-p", args->port, "-r", "-t", args->elf, (char *)NULL);
        perror("exec propeller-load");
        _exit(127);
    }

    close(pipefd[1]);
    int flags = fcntl(pipefd[0], F_GETFL, 0);
    if (flags >= 0) {
        fcntl(pipefd[0], F_SETFL, flags | O_NONBLOCK);
    }

    proc->pid = pid;
    proc->fd = pipefd[0];
    proc->exit_code = -1;
    proc->exited = false;
    proc->stopped_by_parent = false;
    proc->pending_len = 0;
    proc->pending[0] = '\0';
    return 0;
}

static void handle_line(
    const char *raw_line,
    Sample *latest,
    Trace *left_speed,
    Trace *right_speed,
    float *peak_speed,
    int *peak_count,
    LogRing *logs
) {
    char line[LINE_CAP];
    snprintf(line, sizeof(line), "%s", raw_line);
    strip_line(line);
    if (!line[0]) {
        return;
    }

    Sample sample;
    if (parse_sample(line, &sample, monotonic_seconds())) {
        double dt = sample.t - latest->t;
        if (dt < 0.001) {
            dt = 0.001;
        }
        int dl = sample.left - latest->left;
        int dr = sample.right - latest->right;
        if (dl < 0) dl = 0;
        if (dr < 0) dr = 0;

        float ls = (float)((double)dl / dt);
        float rs = (float)((double)dr / dt);
        *peak_speed = fmaxf(fmaxf(*peak_speed * 0.997f, ls), fmaxf(rs, 1.0f));
        if (sample.left > *peak_count) *peak_count = sample.left;
        if (sample.right > *peak_count) *peak_count = sample.right;
        if (*peak_count < 1) *peak_count = 1;
        trace_push(left_speed, ls);
        trace_push(right_speed, rs);
        *latest = sample;
    } else {
        log_push(logs, line);
    }
}

static void pump_loader(
    LoaderProc *proc,
    Sample *latest,
    Trace *left_speed,
    Trace *right_speed,
    float *peak_speed,
    int *peak_count,
    LogRing *logs
) {
    if (proc->fd >= 0) {
        for (;;) {
            char buf[512];
            ssize_t n = read(proc->fd, buf, sizeof(buf));
            if (n > 0) {
                for (ssize_t i = 0; i < n; i++) {
                    char ch = buf[i];
                    if (proc->pending_len < LINE_CAP - 1) {
                        proc->pending[proc->pending_len++] = ch;
                        proc->pending[proc->pending_len] = '\0';
                    }
                    if (ch == '\n') {
                        handle_line(proc->pending, latest, left_speed, right_speed, peak_speed, peak_count, logs);
                        proc->pending_len = 0;
                        proc->pending[0] = '\0';
                    }
                }
                continue;
            }
            if (n == 0) {
                if (proc->pending_len > 0) {
                    handle_line(proc->pending, latest, left_speed, right_speed, peak_speed, peak_count, logs);
                    proc->pending_len = 0;
                    proc->pending[0] = '\0';
                }
                close(proc->fd);
                proc->fd = -1;
                break;
            }
            if (errno == EAGAIN || errno == EWOULDBLOCK) {
                break;
            }
            log_push(logs, strerror(errno));
            close(proc->fd);
            proc->fd = -1;
            break;
        }
    }

    if (!proc->exited && proc->pid > 0) {
        int status = 0;
        pid_t result = waitpid(proc->pid, &status, WNOHANG);
        if (result == proc->pid) {
            proc->exited = true;
            if (WIFEXITED(status)) {
                proc->exit_code = WEXITSTATUS(status);
            } else if (WIFSIGNALED(status)) {
                proc->exit_code = 128 + WTERMSIG(status);
            } else {
                proc->exit_code = status;
            }
        }
    }
}

static void stop_loader(LoaderProc *proc) {
    if (proc->pid <= 0 || proc->exited) {
        if (proc->fd >= 0) {
            close(proc->fd);
            proc->fd = -1;
        }
        return;
    }

    kill(-proc->pid, SIGINT);
    proc->stopped_by_parent = true;
    for (int i = 0; i < 20; i++) {
        int status = 0;
        pid_t result = waitpid(proc->pid, &status, WNOHANG);
        if (result == proc->pid) {
            proc->exited = true;
            if (WIFEXITED(status)) {
                proc->exit_code = WEXITSTATUS(status);
            } else if (WIFSIGNALED(status)) {
                proc->exit_code = 128 + WTERMSIG(status);
            }
            break;
        }
        sleep_ms(100);
    }

    if (!proc->exited) {
        kill(-proc->pid, SIGKILL);
        waitpid(proc->pid, NULL, 0);
        proc->exited = true;
        proc->exit_code = 137;
    }

    if (proc->fd >= 0) {
        close(proc->fd);
        proc->fd = -1;
    }
}

static void write_summary(const Args *args, const Sample *latest, float peak_speed, const LogRing *logs) {
    FILE *fp = fopen(args->summary_file, "w");
    if (!fp) {
        return;
    }
    fprintf(fp, "port=%s\n", args->port);
    fprintf(fp, "board=%s\n", args->board);
    fprintf(fp, "left_edges=%d\n", latest->left);
    fprintf(fp, "right_edges=%d\n", latest->right);
    fprintf(fp, "left_raw=%d\n", latest->left_raw);
    fprintf(fp, "right_raw=%d\n", latest->right_raw);
    fprintf(fp, "peak_edges_per_second=%.2f\n", peak_speed);
    fprintf(fp, "logs:\n");
    for (int i = 0; i < logs->count; i++) {
        fprintf(fp, "%s\n", log_at(logs, i));
    }
    fclose(fp);
}

static void print_usage(const char *argv0) {
    fprintf(stderr,
        "Usage: %s [--port /dev/ttyUSB0] [--board activityboard] [--seconds N]\n"
        "          [--width W] [--height H] [--loader PATH] [--elf PATH]\n"
        "          [--summary-file PATH]\n",
        argv0);
}

static int parse_args(int argc, char **argv, Args *args) {
    static char loader_path[1024];
    static char elf_path[1024];
    static char summary_path[1024];

    path_join(loader_path, sizeof(loader_path), SIM2REAL_ROOT, "tools/parallax/simpleide/opt/parallax/bin/propeller-load");
    path_join(elf_path, sizeof(elf_path), LINE_FOLLOW_ROOT, "build/parallax-smoke/encoder_stream.elf");
    path_join(summary_path, sizeof(summary_path), LINE_FOLLOW_ROOT, "build/parallax-smoke/encoder-dashboard-raylib-c-last.txt");

    args->port = env_default("PROPELLER_LOAD_PORT", "/dev/ttyUSB0");
    args->board = env_default("PROPELLER_LOAD_BOARD", "activityboard");
    args->loader = loader_path;
    args->elf = elf_path;
    args->summary_file = summary_path;
    args->width = 1120;
    args->height = 720;
    args->seconds = 0.0;

    for (int i = 1; i < argc; i++) {
        const char *arg = argv[i];
        if (strcmp(arg, "--port") == 0 && i + 1 < argc) {
            args->port = argv[++i];
        } else if (strcmp(arg, "--board") == 0 && i + 1 < argc) {
            args->board = argv[++i];
        } else if (strcmp(arg, "--loader") == 0 && i + 1 < argc) {
            args->loader = argv[++i];
        } else if (strcmp(arg, "--elf") == 0 && i + 1 < argc) {
            args->elf = argv[++i];
        } else if (strcmp(arg, "--summary-file") == 0 && i + 1 < argc) {
            args->summary_file = argv[++i];
        } else if (strcmp(arg, "--width") == 0 && i + 1 < argc) {
            args->width = atoi(argv[++i]);
        } else if (strcmp(arg, "--height") == 0 && i + 1 < argc) {
            args->height = atoi(argv[++i]);
        } else if (strcmp(arg, "--seconds") == 0 && i + 1 < argc) {
            args->seconds = atof(argv[++i]);
        } else if (strcmp(arg, "--help") == 0 || strcmp(arg, "-h") == 0) {
            print_usage(argv[0]);
            return 1;
        } else {
            fprintf(stderr, "Unknown or incomplete argument: %s\n", arg);
            print_usage(argv[0]);
            return -1;
        }
    }

    if (args->width < 720) args->width = 720;
    if (args->height < 520) args->height = 520;
    return 0;
}

int main(int argc, char **argv) {
    Args args;
    int parsed = parse_args(argc, argv, &args);
    if (parsed != 0) {
        return parsed < 0 ? 2 : 0;
    }

    LogRing logs = {0};
    if (!file_exists(args.loader)) {
        fprintf(stderr, "propeller-load not found: %s\n", args.loader);
        return 2;
    }
    if (!file_exists(args.elf)) {
        fprintf(stderr, "encoder stream ELF not found: %s\n", args.elf);
        fprintf(stderr, "Run ocean/line_follow/scripts/parallax-build-encoder-stream.sh first.\n");
        return 2;
    }

    char cmd_log[LOG_LEN];
    snprintf(cmd_log, sizeof(cmd_log), "propeller-load -b %s -p %s -r -t encoder_stream.elf", args.board, args.port);
    log_push(&logs, cmd_log);

    LoaderProc proc = {.pid = -1, .fd = -1, .exit_code = -1, .exited = false, .stopped_by_parent = false, .pending_len = 0};
    if (start_loader(&proc, &args, &logs) != 0) {
        return 2;
    }

    SetTraceLogLevel(LOG_WARNING);
    InitWindow(args.width, args.height, "ActivityBot Encoder Dashboard C");
    SetTargetFPS(60);

    Trace left_speed = {0};
    Trace right_speed = {0};
    Sample latest = {.t = monotonic_seconds(), .left = 0, .right = 0, .left_raw = 0, .right_raw = 0};
    float peak_speed = 1.0f;
    int peak_count = 1;
    double start = monotonic_seconds();

    while (!WindowShouldClose()) {
        pump_loader(&proc, &latest, &left_speed, &right_speed, &peak_speed, &peak_count, &logs);
        if (IsKeyPressed(KEY_Q)) {
            break;
        }
        if (args.seconds > 0.0 && monotonic_seconds() - start >= args.seconds) {
            break;
        }

        int width = GetScreenWidth();
        int height = GetScreenHeight();
        int graph_w = width - 120;
        if (graph_w < 400) graph_w = 400;

        BeginDrawing();
        ClearBackground(C_BG);

        DrawRectangle(28, 24, width - 56, 96, C_PANEL);
        draw_label(48, 42, "ActivityBot Encoder Dashboard", 32, RAYWHITE);

        char line[256];
        snprintf(line, sizeof(line), "%s  |  RAM-only  |  spin wheels by hand  |  Q/Esc closes", args.port);
        draw_label(50, 82, line, 20, C_MUTED);

        snprintf(line, sizeof(line), "LEFT  edges %6d   raw P14=%d   speed %6.1f/s", latest.left, latest.left_raw, trace_recent(&left_speed));
        draw_label(48, 148, line, 24, C_CYAN);
        draw_bar(48, 184, graph_w, 34, (float)latest.left, (float)peak_count, C_CYAN);

        snprintf(line, sizeof(line), "RIGHT edges %6d   raw P15=%d   speed %6.1f/s", latest.right, latest.right_raw, trace_recent(&right_speed));
        draw_label(48, 250, line, 24, C_ORANGE);
        draw_bar(48, 286, graph_w, 34, (float)latest.right, (float)peak_count, C_ORANGE);

        snprintf(line, sizeof(line), "Scrolling speed trace   peak scale %.1f edges/s", peak_speed);
        draw_label(48, 356, line, 22, RAYWHITE);
        draw_trace(48, 390, graph_w, 86, &left_speed, peak_speed, C_CYAN);
        draw_trace(48, 500, graph_w, 86, &right_speed, peak_speed, C_ORANGE);
        draw_label(60, 396, "LEFT", 18, C_CYAN);
        draw_label(60, 506, "RIGHT", 18, C_ORANGE);

        DrawCircle(width - 54, 160, 11, latest.left_raw ? C_GREEN : C_OFF);
        DrawCircle(width - 54, 262, 11, latest.right_raw ? C_GREEN : C_OFF);

        int log_y = height - 140;
        draw_label(48, log_y, "loader / serial", 18, C_MUTED);
        int visible_logs = logs.count < 5 ? logs.count : 5;
        int first_log = logs.count - visible_logs;
        for (int i = 0; i < visible_logs; i++) {
            draw_label(48, log_y + 24 + i * 20, log_at(&logs, first_log + i), 18, C_MUTED);
        }

        if (proc.exited) {
            snprintf(line, sizeof(line), "propeller-load exited: %d", proc.exit_code);
            draw_label(48, height - 36, line, 18, RED);
        } else {
            draw_label(48, height - 36, "live from P14/P15 through usbipd -> /dev/ttyUSB0 -> propeller-load terminal mode", 18, C_MUTED);
        }

        EndDrawing();
    }

    stop_loader(&proc);
    CloseWindow();
    write_summary(&args, &latest, peak_speed, &logs);
    if (proc.stopped_by_parent) {
        return 0;
    }
    return (proc.exit_code == -1 || proc.exit_code == 0 || proc.exit_code == 130) ? 0 : proc.exit_code;
}

File: scripts/activitybot-encoder-dashboard-raylib.py
#!/usr/bin/env python3
"""Raylib live dashboard for ActivityBot encoder edges."""

from __future__ import annotations

import argparse
import collections
import os
import queue
import re
import signal
import subprocess
import sys
import threading
import time
from pathlib import Path

import pyray as rl


STREAM_RE = re.compile(r"^E\s+(-?\d+)\s+(-?\d+)\s+([01])\s+([01])")


class Sample:
    def __init__(self, t: float, left: int, right: int, left_raw: int, right_raw: int):
        self.t = t
        self.left = left
        self.right = right
        self.left_raw = left_raw
        self.right_raw = right_raw


def repo_root() -> Path:
    return Path(__file__).resolve().parents[1]


def build_firmware(root: Path) -> list[str]:
    result = subprocess.run(
        [str(root / "scripts/parallax-build-encoder-stream.sh")],
        check=True,
        capture_output=True,
        text=True,
    )
    return [line for line in result.stdout.splitlines() if line.strip()]


def read_loader(proc: subprocess.Popen[str], out: queue.Queue[tuple[str, object]]) -> None:
    assert proc.stdout is not None
    for line in proc.stdout:
        clean = line.replace("\r", "").strip()
        match = STREAM_RE.search(clean)
        if match:
            out.put((
                "sample",
                Sample(
                    time.monotonic(),
                    int(match.group(1)),
                    int(match.group(2)),
                    int(match.group(3)),
                    int(match.group(4)),
                ),
            ))
        elif clean:
            out.put(("log", clean))
    out.put(("exit", proc.poll()))


def draw_label(x: int, y: int, text: str, size: int = 22, color=None) -> None:
    rl.draw_text(text, x, y, size, color or rl.RAYWHITE)


def draw_bar(x: int, y: int, w: int, h: int, value: float, peak: float, color) -> None:
    rl.draw_rectangle(x, y, w, h, rl.Color(36, 39, 46, 255))
    fill = 0 if peak <= 0 else int(max(0, min(w, (value / peak) * w)))
    rl.draw_rectangle(x, y, fill, h, color)
    rl.draw_rectangle_lines(x, y, w, h, rl.Color(120, 126, 138, 255))


def draw_trace(x: int, y: int, w: int, h: int, values: collections.deque[float], peak: float, color) -> None:
    rl.draw_rectangle(x, y, w, h, rl.Color(22, 25, 31, 255))
    rl.draw_rectangle_lines(x, y, w, h, rl.Color(80, 86, 96, 255))
    rl.draw_line(x, y + h // 2, x + w, y + h // 2, rl.Color(45, 50, 58, 255))
    recent = list(values)[-w:]
    if len(recent) < 2:
        return
    start_x = x + max(0, w - len(recent))
    scale = max(peak, 1.0)
    prev_x = start_x
    prev_y = y + h - int((recent[0] / scale) * (h - 8)) - 4
    for idx, value in enumerate(recent[1:], start=1):
        px = start_x + idx
        py = y + h - int((value / scale) * (h - 8)) - 4
        py = max(y + 3, min(y + h - 3, py))
        rl.draw_line(prev_x, prev_y, px, py, color)
        prev_x, prev_y = px, py


def run(args: argparse.Namespace) -> int:
    root = repo_root()
    if args.build:
        build_logs = build_firmware(root)
    else:
        build_logs = []

    loader = root / "tools/parallax/simpleide/opt/parallax/bin/propeller-load"
    elf = root / "build/parallax-smoke/encoder_stream.elf"
    cmd = [
        str(loader),
        "-b",
        args.board,
        "-p",
        args.port,
        "-r",
        "-t",
        str(elf),
    ]

    proc = subprocess.Popen(
        cmd,
        cwd=str(root),
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
        errors="replace",
        bufsize=1,
    )
    events: queue.Queue[tuple[str, object]] = queue.Queue()
    threading.Thread(target=read_loader, args=(proc, events), daemon=True).start()

    rl.init_window(args.width, args.height, "ActivityBot Encoder Dashboard")
    rl.set_target_fps(60)

    logs: collections.deque[str] = collections.deque(build_logs[-4:], maxlen=7)
    left_speed: collections.deque[float] = collections.deque(maxlen=max(120, args.width - 160))
    right_speed: collections.deque[float] = collections.deque(maxlen=max(120, args.width - 160))
    latest = Sample(time.monotonic(), 0, 0, 0, 0)
    peak_speed = 1.0
    peak_count = 1
    start = time.monotonic()
    exit_code: int | None = None

    bg = rl.Color(15, 18, 24, 255)
    panel = rl.Color(27, 31, 39, 255)
    cyan = rl.Color(64, 196, 255, 255)
    orange = rl.Color(255, 176, 64, 255)
    green = rl.Color(94, 220, 126, 255)
    muted = rl.Color(155, 164, 178, 255)

    try:
        while not rl.window_should_close():
            while True:
                try:
                    kind, payload = events.get_nowait()
                except queue.Empty:
                    break
                if kind == "sample":
                    sample = payload
                    assert isinstance(sample, Sample)
                    dt = max(1e-3, sample.t - latest.t)
                    dl = max(0, sample.left - latest.left)
                    dr = max(0, sample.right - latest.right)
                    ls = dl / dt
                    rs = dr / dt
                    peak_speed = max(peak_speed * 0.997, ls, rs, 1.0)
                    peak_count = max(peak_count, sample.left, sample.right, 1)
                    left_speed.append(ls)
                    right_speed.append(rs)
                    latest = sample
                elif kind == "log":
                    logs.append(str(payload))
                elif kind == "exit":
                    exit_code = payload if isinstance(payload, int) else proc.poll()

            if rl.is_key_pressed(rl.KeyboardKey.KEY_Q) or rl.is_key_pressed(rl.KeyboardKey.KEY_ESCAPE):
                break
            if args.seconds and time.monotonic() - start >= args.seconds:
                break

            width = rl.get_screen_width()
            height = rl.get_screen_height()
            graph_w = width - 120

            rl.begin_drawing()
            rl.clear_background(bg)

            rl.draw_rectangle(28, 24, width - 56, 96, panel)
            draw_label(48, 42, "ActivityBot Encoder Dashboard", 32, rl.RAYWHITE)
            draw_label(50, 82, f"{args.port}  |  RAM-only  |  spin wheels by hand  |  Q/Esc closes", 20, muted)

            draw_label(48, 148, f"LEFT  edges {latest.left:6d}   raw P14={latest.left_raw}   speed {left_speed[-1] if left_speed else 0:6.1f}/s", 24, cyan)
            draw_bar(48, 184, graph_w, 34, latest.left, peak_count, cyan)

            draw_label(48, 250, f"RIGHT edges {latest.right:6d}   raw P15={latest.right_raw}   speed {right_speed[-1] if right_speed else 0:6.1f}/s", 24, orange)
            draw_bar(48, 286, graph_w, 34, latest.right, peak_count, orange)

            draw_label(48, 356, f"Scrolling speed trace   peak scale {peak_speed:0.1f} edges/s", 22, rl.RAYWHITE)
            draw_trace(48, 390, graph_w, 86, left_speed, peak_speed, cyan)
            draw_trace(48, 500, graph_w, 86, right_speed, peak_speed, orange)
            draw_label(60, 396, "LEFT", 18, cyan)
            draw_label(60, 506, "RIGHT", 18, orange)

            raw_y = 148
            rl.draw_circle(width - 54, raw_y + 12, 11, green if latest.left_raw else rl.Color(70, 75, 84, 255))
            rl.draw_circle(width - 54, raw_y + 114, 11, green if latest.right_raw else rl.Color(70, 75, 84, 255))

            log_y = height - 140
            draw_label(48, log_y, "loader / serial", 18, muted)
            for idx, line in enumerate(list(logs)[-5:]):
                draw_label(48, log_y + 24 + idx * 20, line[:110], 18, muted)

            if exit_code is not None:
                draw_label(48, height - 36, f"propeller-load exited: {exit_code}", 18, rl.RED)
            else:
                draw_label(48, height - 36, "live from P14/P15 through usbipd -> /dev/ttyUSB0 -> propeller-load terminal mode", 18, muted)

            rl.end_drawing()
    finally:
        if proc.poll() is None:
            proc.send_signal(signal.SIGINT)
            try:
                proc.wait(timeout=2)
            except subprocess.TimeoutExpired:
                proc.kill()
                proc.wait(timeout=2)
        rl.close_window()

    if args.summary_file:
        summary = Path(args.summary_file)
        if not summary.is_absolute():
            summary = root / summary
        summary.parent.mkdir(parents=True, exist_ok=True)
        summary.write_text(
            "\n".join(
                [
                    f"port={args.port}",
                    f"board={args.board}",
                    f"left_edges={latest.left}",
                    f"right_edges={latest.right}",
                    f"left_raw={latest.left_raw}",
                    f"right_raw={latest.right_raw}",
                    f"peak_edges_per_second={peak_speed:.2f}",
                    "logs:",
                    *list(logs),
                ]
            )
            + "\n"
        )

    return 0 if exit_code in (None, 0) else int(exit_code)


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--port", default=os.environ.get("PROPELLER_LOAD_PORT", "/dev/ttyUSB0"))
    parser.add_argument("--board", default=os.environ.get("PROPELLER_LOAD_BOARD", "activityboard"))
    parser.add_argument("--width", type=int, default=1120)
    parser.add_argument("--height", type=int, default=720)
    parser.add_argument("--seconds", type=float, default=0.0)
    parser.add_argument("--no-build", dest="build", action="store_false")
    parser.add_argument("--summary-file", default="build/parallax-smoke/encoder-dashboard-raylib-last.txt")
    parser.set_defaults(build=True)
    return parser.parse_args(argv)


def main(argv: list[str]) -> int:
    return run(parse_args(argv))


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
